crypto/sm2/sm2_pmeth.c

```diff
@@ -0,0 +1,269 @@
+/*
+ * Copyright 2017 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright 2017 Ribose Inc. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/sm2.h>
+#include "internal/cryptlib.h"
+#include <openssl/asn1t.h>
+#include <openssl/x509.h>
+#include <openssl/ec.h>
+#include <openssl/evp.h>
+#include <openssl/sm2err.h>
+#include "internal/evp_int.h"
+
+/* EC pkey context structure */
+
+typedef struct {
+    EC_GROUP *gen_group;
+    const EVP_MD *md;
+    char* user_id;
+} SM2_PKEY_CTX;
+
+static int pkey_sm2_init(EVP_PKEY_CTX *ctx)
+{
+    SM2_PKEY_CTX *dctx;
+
+    dctx = OPENSSL_zalloc(sizeof(*dctx));
+    if (dctx == NULL)
+        return 0;
+
+    ctx->data = dctx;
+    return 1;
+}
+
+static int pkey_sm2_copy(EVP_PKEY_CTX *dst, EVP_PKEY_CTX *src)
+{
+    SM2_PKEY_CTX *dctx, *sctx;
+    if (!pkey_sm2_init(dst))
+        return 0;
+    sctx = src->data;
+    dctx = dst->data;
+    if (sctx->gen_group) {
+        dctx->gen_group = EC_GROUP_dup(sctx->gen_group);
+        if (!dctx->gen_group)
+            return 0;
+    }
+    dctx->md = sctx->md;
+    return 1;
+}
+
+static void pkey_sm2_cleanup(EVP_PKEY_CTX *ctx)
+{
+    SM2_PKEY_CTX *dctx = ctx->data;
+    if (dctx) {
+        EC_GROUP_free(dctx->gen_group);
+        OPENSSL_free(dctx);
+    }
+}
+
+static int pkey_sm2_sign(EVP_PKEY_CTX *ctx, unsigned char *sig, size_t *siglen,
+                        const unsigned char *tbs, size_t tbslen)
+{
+    SM2_PKEY_CTX *dctx = ctx->data;
+    EC_KEY *ec = ctx->pkey->pkey.ec;
+    ECDSA_SIG* der_sig = NULL;
+    const EVP_MD* md = NULL;
+    const char* user_id = NULL;
+
+    if (!sig) {
+        /* ECDSA and SM2 signatures have the same size */
+        *siglen = ECDSA_size(ec);
+        return 1;
+    } else if (*siglen < (size_t)ECDSA_size(ec)) {
+        ECerr(EC_F_PKEY_SM2_SIGN, SM2_R_BUFFER_TOO_SMALL);
+        return 0;
+    }
+
+    md = (dctx->md) ? dctx->md : EVP_sm3();
+
+    user_id = (dctx->user_id) ? dctx->user_id : SM2_DEFAULT_USERID;
+
+    der_sig = SM2_do_sign(ec, md, user_id, tbs, tbslen);
+
+    if(der_sig == NULL)
+       return 0;
+
+    /* Now ASN.1 encode ... */
+
+    *siglen = i2d_ECDSA_SIG(der_sig, &sig);
+
+    return 1;
+}
+
+static int pkey_sm2_verify(EVP_PKEY_CTX *ctx,
+                          const unsigned char *sig, size_t siglen,
+                          const unsigned char *tbs, size_t tbslen)
+{
+    int ret, type;
+    SM2_PKEY_CTX *dctx = ctx->data;
+    EC_KEY *ec = ctx->pkey->pkey.ec;
+
+    if (dctx->md)
+        type = EVP_MD_type(dctx->md);
+    else
+        type = NID_sm3;
+
+    ret = ECDSA_verify(type, tbs, tbslen, sig, siglen, ec);
```

Comment:
Am I missing something or should this not be calling SM2_do_verify() rather than ECDSA_verify()?

Reply:
Actually I'm wondering if the issue is that the sign is incorrect in calling SM2_do_sign() directly.
My interest in this pull request is to have an SM2 implementation that can be offloaded to an engine (the same as ECDSA). If I'm reading it correctly, the way pkey_sm2_sign() is written now it calls the SW implementation of SM2_do_sign() directly, there is no function pointer eckey->method->sign used to select the implementation allowing an engine to provide it's own version of the code. I would have expected an intermediate function like ECDSA provides: ECDSA_sign() (perhaps SM2_sign() that calls the function pointer) and that be called from pkey_sm2_sign(). I would expect a similar implementation for the verify as well.
I'm guessing this is all down to the issue you raised about ECKEY's being hardcoded for ECDSA and how you get it to work with SM2.

Reply:
Ok I've delved a bit further, maybe my thinking was not correct. I was assuming an engine would offload by providing its own EC_KEY_METHOD and tie in at the lower level like I currently do to provide an engine implementation of ECDSA. Looking at the current implementation of X25519/ed25519 I believe the expectation is for an engine to provide its own EVP_PKEY_METHOD. If that is the case then perhaps my initial comment stands that here it should be calling SM2_do_verify() rather than ECDSA_verify().

Reply:
This part of the code is just incomplete. But it turns out that instead we should integrate SM2 with the existing `ec_pmeth.c` and then this file goes away entirely. I'm working on that now.